server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java

```diff
@@ -0,0 +1,424 @@
+                    + "load is not available",
+                dataStream.getName()
+            );
+            return NOT_APPLICABLE_RESULT;
+        }
+        return innerCalculate(metadata, dataStream, writeIndexLoad, nowSupplier);
+    }
+
+    private AutoShardingResult innerCalculate(Metadata metadata, DataStream dataStream, double writeIndexLoad, LongSupplier nowSupplier) {
+        // increasing the number of shards is calculated solely based on the index load of the write index
+        IndexMetadata writeIndex = metadata.index(dataStream.getWriteIndex());
+        assert writeIndex != null : "the data stream write index must exist in the provided cluster metadata";
+        AutoShardingResult increaseShardsResult = getIncreaseShardsResult(dataStream, writeIndexLoad, nowSupplier, writeIndex);
+        if (increaseShardsResult != null) {
+            return increaseShardsResult;
+        }
+
+        // the way we handle the cool down when reducing the number of shards is concerned is a bit different as it involves checking the
+        // write load of more than just the write index.
+        // because we have to look at the maximum write load encountered within the configured cool down period this can involve
+        // checking the write load for older backing indices. if the cool down period hasn't lapsed we short circuit here and return
+        // NO_CHANGE_REQUIRED because we don't want our historic write load lookup to include backing indices that are older than the
+        // index that triggered the previous auto scaling event
+        // this would happen if say, we increased the number of shards one day ago, and there's 2 days left of the reduce number of shards
+        // cool down - looking back the entire cool down period (3 days) would include the previous couple backing indices, including the
+        // one before the index that triggered the increase number of shards event)
+        TimeValue remainingReduceShardsCooldown = getRemainingDecreaseShardsCooldown(metadata, dataStream);
+        if (remainingReduceShardsCooldown.equals(TimeValue.ZERO) == false) {
+            Index oldestBackingIndex = dataStream.getIndices().get(0);
+            IndexMetadata oldestIndexMeta = metadata.getIndexSafe(oldestBackingIndex);
+            logger.debug(
+                "data stream autosharding service recommends no change in the number of shards for data stream [{}]  "
+                    + "with previous autosharding event [{}] and oldest generation index [{}] being created on [{}] ",
+                dataStream.getName(),
+                dataStream.getAutoShardingEvent(),
+                oldestBackingIndex.getName(),
+                oldestIndexMeta.getCreationDate()
+            );
+
+            return new AutoShardingResult(
+                AutoShardingType.NO_CHANGE_REQUIRED,
+                writeIndex.getNumberOfShards(),
+                writeIndex.getNumberOfShards(),
+                TimeValue.ZERO,
```

Comment:
It seems strange to me that we return 0 for the remaining cooldown, but we actually still have a cooldown time for the decreasing.

Do you think instead, maybe we should change the enum to have things like `COOLDOWN_PREVENTED_DECREASE` and `COOLDOWN_PREVENTED_INCREASE` to indicate this exactly, differently than `NO_CHANGE_REQUIRED`? (because we didn't actually compute whether we *would* have decreased in this case)

Reply:
The nice thing about this is that then we don't have to rely on the caller checking the cooldown remaining for `INCREASE/DECREASE`, instead, they can do a `switch` statement on the enum value, and see either `INCREASE_SHARDS`, `DECREASE_SHARDS`, `NO_CHANGE_REQUESTED`, `NOT_APPLICABLE`, `COOLDOWN_PREVENTED_DECREASE`, or `COOLDOWN_PREVENTED_INCREASE` and make the appropriate logging/decision. The time remaining becomes more of an informative thing instead of a decision-making thing.

Reply:
I also like this because then we could return "oh here's what the number of shards _WOULD BE_ if you increased/decreased right now", which can help us in the future debug whether we need to tune the cooldown periods.

Reply:
I've added this suggestion and it's looking pretty neat. Thanks ! 

Love that the outside world doesn't have to care about the cooldown value. 

Note that now the `NO_CHANGE_REQUIRED` continues to have a cooldown of zero (which IMO makes sense) as otherwise one of the other statuses is returned.